sklearn/metrics/pairwise.py

```diff
@@ -829,6 +833,177 @@ def cosine_distances(X, Y=None):
     return S
 
 
+def _split_categorical_numerical(X, categorical_features):
+    # the following bit is done before check_pairwise_array to avoid converting
+    # numerical data to object dtype. First we split the data into categorical
+    # and numerical, then we do check_array
+
+    # TODO: this should be more like check_array(..., accept_pandas=True)
+    if (X is not None and not hasattr(X, 'iloc')
+            and not hasattr(X, '__array__')):
+        X = check_array(X, dtype=np.object, force_all_finite=False)
+
+    if callable(categorical_features) and X is not None:
+        cols = categorical_features(X)
+    else:
+        cols = categorical_features
+    if cols is None:
+        cols = []
+
+    if X is not None:
+        X_cat = _safe_indexing(X, cols, axis=1)
+        X_num = _safe_indexing(X, cols, axis=1, inverse=True)
+    else:
+        X_cat = X_num = None
+    return X_cat, X_num
+
+
+def gower_distances(X, Y=None, categorical_features=None, scale=True,
+                    min_values=None, scale_factor=None):
+    """Compute the distances between the observations in X and Y,
+    that may contain mixed types of data, using an implementation
+    of Gower formula.
+
+    Parameters
+    ----------
+    X : {array-like, pandas.DataFrame} of shape (n_samples, n_features)
+
+    Y : {array-like, pandas.DataFrame} of shape (n_samples, n_features), \
+        default=None
+
+    categorical_features : array-like of str, array-like of int, \
+            array-like of bool, slice or callable, default=None
+        Indexes the data on its second axis. Integers are interpreted as
+        positional columns, while strings can reference DataFrame columns
+        by name.
+        A callable is passed the input data `X` and can return any of the
+        above. To select multiple columns by name or dtype, you can use
+        :obj:`make_column_selector`.
```

Comment:
Can't check the rendered docs yet (need to update classes.rst) but I think this should be ``` :func:`~sklearn.compose.make_column_selector` ```

Reply:
I was wondering why this works, but it's a copy/paste from `ColumnTransformer`.

Reply:
`obj` might work but I think that not prefixing with `sklearn.compose` will not link. It worked before because these 2 were in the same file